insonmnia/miner/network/l2tp_state.go

```diff
@@ -0,0 +1,267 @@
+	boltdb.Register()
+	var (
+		backend   = store.Backend(store.BOLTDB)
+		endpoints = []string{path}
+		config    = store.Config{Bucket: "sonm_l2tp_driver_state"}
+	)
+	storage, err := libkv.NewStore(backend, endpoints, &config)
+	if err != nil {
+		return nil, err
+	}
+
+	state := &l2tpState{
+		Aliases:  make(map[string]string),
+		Networks: make(map[string]*l2tpNetwork),
+		logger:   log.S(ctx).With("source", "l2tp/store"),
+		storage:  storage,
+	}
+
+	if err := state.load(); err != nil {
+		return nil, err
+	}
+
+	return state, nil
+}
+
+func (s *l2tpState) Sync() (err error) {
+	defer func() {
+		if err != nil {
+			s.logger.Error("failed to sync l2tp state", zap.Error(err))
+		}
+	}()
+
+	marshaled, err := json.Marshal(s)
+	if err != nil {
+		return err
+	}
+
+	err = s.storage.Put("state", marshaled, &store.WriteOptions{})
+	return
+}
+
+func (s *l2tpState) AddNetwork(netID string, netInfo *l2tpNetwork) error {
+	if _, ok := s.Aliases[netID]; ok {
+		return errors.Errorf("network already exists: %s", netID)
+	}
+
+	s.Aliases[netID] = netID
+	s.Networks[netID] = netInfo
+
+	return nil
+}
+
+func (s *l2tpState) AddNetworkAlias(netID, alias string) error {
+	if _, ok := s.Aliases[netID]; !ok {
+		return errors.Errorf("network not found: %s", netID)
+	}
+
+	s.Aliases[alias] = netID
+
+	return nil
+}
+
+func (s *l2tpState) RemoveNetwork(netID string) error {
+	translatedID, ok := s.Aliases[netID]
+	if !ok {
+		return errors.Errorf("network not found: %s", netID)
+	}
+
+	delete(s.Networks, translatedID)
+
+	for alias, target := range s.Aliases {
+		if target == translatedID {
+			delete(s.Aliases, alias)
+		}
+	}
+
+	return nil
+}
+
+func (s *l2tpState) GetNetwork(netID string) (*l2tpNetwork, error) {
+	translatedID, ok := s.Aliases[netID]
+	if !ok {
+		return nil, errors.Errorf("network not found: %s", netID)
+	}
+
+	if netInfo, ok := s.Networks[translatedID]; ok {
+		return netInfo, nil
+	}
+
+	return nil, errors.Errorf("network not found: %s", netID)
+}
+
+type l2tpNetwork struct {
+	ID           string
+	PoolID       string
+	Count        int
+	NetworkOpts  *l2tpNetworkConfig
+	Endpoint     *l2tpEndpoint
+	NeedsGateway bool
+}
+
+func newL2tpNetwork(opts *l2tpNetworkConfig) *l2tpNetwork {
+	return &l2tpNetwork{
+		NetworkOpts:  opts,
+		NeedsGateway: true,
+	}
+}
+
+func (n *l2tpNetwork) Setup() error {
+	n.PoolID = n.NetworkOpts.GetPoolID()
+
+	return nil
+}
+
+func (n *l2tpNetwork) ConnInc() {
+	n.Count++
+}
+
+type l2tpEndpoint struct {
+	ID           string
+	Name         string
+	ConnName     string
+	PPPOptFile   string
+	PPPDevName   string
+	AssignedCIDR string
+	AssignedIP   string
+	NetworkOpts  *l2tpNetworkConfig
+}
+
+func NewL2TPEndpoint(netInfo *l2tpNetwork) *l2tpEndpoint {
+	return &l2tpEndpoint{
+		Name:        fmt.Sprintf("%d_%s", netInfo.Count, netInfo.PoolID),
+		NetworkOpts: netInfo.NetworkOpts,
+	}
+}
+
+func (e *l2tpEndpoint) setup() error {
+	e.ConnName = e.Name + "-connection"
+	e.PPPDevName = ("ppp" + e.Name)[:15]
```

Comment:
Are you sure that the name will contain at least 12-bytes?

Reply:
absolutely, it's based on an md5 hash